src/IO/WriteBufferFromS3.cpp

```diff
@@ -217,30 +242,27 @@ void WriteBufferFromS3::createMultipartUpload()
         throw S3Exception(outcome.GetError().GetMessage(), outcome.GetError().GetErrorType());
 }
 
-void WriteBufferFromS3::writePart()
+void WriteBufferFromS3::writePart([[maybe_unused]] bool last_part)
 {
-    auto size = temporary_buffer->tellp();
-
-    LOG_TRACE(log, "Writing part. Bucket: {}, Key: {}, Upload_id: {}, Size: {}", bucket, key, multipart_upload_id, size);
-
-    if (size < 0)
-    {
-        LOG_WARNING(log, "Skipping part upload. Buffer is in bad state, it means that we have tried to upload something, but got an exception.");
-        return;
-    }
+    size_t size = offset();
 
     if (size == 0)
     {
-        LOG_TRACE(log, "Skipping writing part. Buffer is empty.");
+        LOG_TRACE(log, "Skipping writing part. Buffer is empty. {}, {}, {}", size, offset(), memory.size());
         return;
     }
 
+    LOG_TRACE(log, "Writing part. Bucket: {}, Key: {}, Upload_id: {}, Size: {}", bucket, key, multipart_upload_id, size);
+
     if (TSA_SUPPRESS_WARNING_FOR_READ(part_tags).size() == S3_WARN_MAX_PARTS)
     {
         // Don't throw exception here by ourselves but leave the decision to take by S3 server.
         LOG_WARNING(log, "Maximum part number in S3 protocol has reached (too many parts). Server may not accept this whole upload.");
     }
 
+    chassert(size <= settings.max_upload_part_size);
+    chassert(size >= settings.min_upload_part_size || last_part);
```

Comment:
I think it's better to throw logical error here.
This check is cheap and better to detect in release build as well.

Reply:
I reworked that part.